### PR TITLE
Set permission fields on initialization rather than request time

### PR DESCRIPTION
**Pull request by gabisurita**
Related to #1033

r? @glasserc @leplatrem

### CODE REVIEW

File: kinto/core/resource/schema.py [hunk 1]
```diff
@@ -74,34 +74,38 @@ class PermissionsSchema(colander.SchemaNode):
...
 
-            # Add a String list child node with the name of ``perm``.
-            self.add(self._get_node_principals(perm))
+        if cstruct in (colander.null, colander.drop):
+            return super(PermissionsSchema, self).deserialize(cstruct)
 
-        # End up by deserializing a mapping whose keys are now known.
-        return super(PermissionsSchema, self).deserialize(permissions)
+        # If permissions are listed, check fields and produce fancy error messages
+        elif self.known_perms:
```
Comment (leplatrem) [inline kinto/core/resource/schema.py]: superfluous `elif` 'cuz of the return above

File: kinto/core/resource/schema.py [hunk 2]
```diff
@@ -74,34 +74,38 @@ class PermissionsSchema(colander.SchemaNode):
...
+            for perm in cstruct:
+                colander.OneOf(choices=self.known_perms)(self, perm)
```
Comment (leplatrem) [inline kinto/core/resource/schema.py]: Why is this necessary if we have `unknown: raise` ?

File: kinto/core/resource/schema.py [hunk 3]
```diff
@@ -74,34 +74,38 @@ class PermissionsSchema(colander.SchemaNode):
...
+            return super(PermissionsSchema, self).deserialize(cstruct)
+
+        # Else deserialize the fields that are not on the schema
+        permissions = {}
+        for perm, principals in cstruct.items():
+            node = colander.SequenceSchema(colander.SchemaNode(colander.String()),
+                                           missing=colander.drop).deserialize(principals)
```
Comment (leplatrem) [inline kinto/core/resource/schema.py]: Why `missing: drop` ?

It's better if you instantiate the schema once out of the loop IMO

File: kinto/core/resource/schema.py [hunk 4]
```diff
@@ -74,34 +74,38 @@ class PermissionsSchema(colander.SchemaNode):
...
-            # Add a String list child node with the name of ``perm``.
-            self.add(self._get_node_principals(perm))
+        if cstruct in (colander.null, colander.drop):
+            return super(PermissionsSchema, self).deserialize(cstruct)
 
-        # End up by deserializing a mapping whose keys are now known.
-        return super(PermissionsSchema, self).deserialize(permissions)
+        # If permissions are listed, check fields and produce fancy error messages
+        elif self.known_perms:
+            for perm in cstruct:
```
Comment (leplatrem) [inline kinto/core/resource/schema.py]: Don't we need to check that `cstruct` is iterable?

File: kinto/core/resource/schema.py [hunk 5]
```diff
@@ -74,34 +74,38 @@ class PermissionsSchema(colander.SchemaNode):
...
+                colander.OneOf(choices=self.known_perms)(self, perm)
+            return super(PermissionsSchema, self).deserialize(cstruct)
+
+        # Else deserialize the fields that are not on the schema
+        permissions = {}
+        for perm, principals in cstruct.items():
```
Comment (leplatrem) [inline kinto/core/resource/schema.py]: Are we sure here that `cstruct` is a mapping?

File: kinto/core/resource/schema.py (hunk 2 above)
Comment (gabisurita) [inline kinto/core/resource/schema.py]: This preserve the error message used previously. e.g:
"permissions in body: \"foo\" is not one of read, write, collection:create, group:create"

Without it we just get an unrecognized fields message. We can set the error massage manually, but then this can override other validation messages.

File: kinto/core/resource/schema.py (hunk 4 above)
Comment (gabisurita) [inline kinto/core/resource/schema.py]: Yes, we do. I should probably add a test for that too.
